### PR TITLE
Refactor BuildMode into class, add jit_release configuration

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

As documented in https://github.com/flutter/flutter/wiki/JIT-Release-Modes. A jit release build uses dill artifacts but with slightly different configuration. Begin support for it in the tool by refactoring build modes.

### CODE REVIEW

File: packages/flutter_tools/lib/src/build_info.dart [hunk 1]
```diff
@@ -139,6 +141,8 @@ BuildMode getBuildModeForName(String name) {
       return BuildMode.profile;
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/build_info.dart]:
```dart
BuildMode getBuildModeForName(String name) {
  return const <String, BuildMode>{
    'debug': BuildMode.debug,
    'profile': BuildMode.profile,
    'release': BuildMode.release,
    'jit_release': BuildMode.jitRelease,
  }[name];
}
```

File: packages/flutter_tools/lib/src/build_system/targets/dart.dart [hunk 2]
```diff
@@ -102,7 +102,7 @@ class CopyFlutterBundle extends Target {
     environment.outputDir.createSync(recursive: true);
 
     // Only copy the prebuilt runtimes and kernel blob in debug mode.
-    if (buildMode == BuildMode.debug) {
+    if (buildMode == BuildMode.debug || buildMode == BuildMode.jitRelease) {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/build_system/targets/dart.dart]: Consider elevating `BuildMode` from an enum to a class (or adding a class that has the enum as a field), and moving the logic here and below to that class.

**Review by zanderso — COMMENTED** [review]
What is the failure mode if you specify a jit_release build targeting iOS?

Do the .dill files produced by the jit_release build use ASTs or bytecode?

**Comment by jonahwilliams** [comment]
> What is the failure mode if you specify a jit_release build targeting iOS?

You won't be able to compile a local engine on iOS, so I didn't add any failure mode here


> Do the .dill files produced by the jit_release build use ASTs or bytecode?

These dill files use AST, but could probably use bytecode - I haven't tested it.

**Review by zanderso — APPROVED** [review]

**Comment by amreniouinnovent** [comment]
@jonahwilliams I wonder building the whole engine is the only way for Android x86 or there will be a simple solution in the future?

**Pull request merged**